pkg/filter/filter_test.go

```diff
@@ -0,0 +1,422 @@
+			expectedResult: false,
+		},
+		{
+			name:           "MachinePool with release label is not processed",
+			inputCR:        machinePool().withReleaseVersionLabel(capiRelease),
+			expectedResult: false,
+		},
+		{
+			name:           "MachinePool with cluster name label is not processed",
+			inputCR:        machinePool().withClusterNameLabel(),
+			ownerCluster:   cluster().withReleaseVersionLabel(capiRelease).object,
+			expectedResult: false,
+		},
+		{
+			name:           "MachinePool with cluster ID label is not processed",
+			inputCR:        machinePool().withClusterIDLabel(),
+			ownerCluster:   cluster().withReleaseVersionLabel(capiRelease).object,
+			expectedResult: false,
+		},
+		// AzureMachinePool
+		{
+			name:           "AzureMachinePool without release label and without cluster name/id label is not processed",
+			inputCR:        azureMachinePool(),
+			expectedResult: false,
+		},
+		{
+			name:           "AzureMachinePool with release label is not processed",
+			inputCR:        azureMachinePool().withReleaseVersionLabel(capiRelease),
+			expectedResult: false,
+		},
+		{
+			name:           "AzureMachinePool with cluster name label is not processed",
+			inputCR:        azureMachinePool().withClusterNameLabel(),
+			ownerCluster:   cluster().withReleaseVersionLabel(capiRelease).object,
+			expectedResult: false,
+		},
+		{
+			name:           "AzureMachinePool with cluster ID label is not processed",
+			inputCR:        azureMachinePool().withClusterIDLabel(),
+			ownerCluster:   cluster().withReleaseVersionLabel(capiRelease).object,
+			expectedResult: false,
+		},
+		// AzureMachine
+		{
+			name:           "AzureMachine without release label and without cluster name/id label is not processed",
+			inputCR:        azureMachine(),
+			expectedResult: false,
+		},
+		{
+			name:           "AzureMachine with release label is not processed",
+			inputCR:        azureMachine().withReleaseVersionLabel(capiRelease),
+			expectedResult: false,
+		},
+		{
+			name:           "AzureMachine with cluster name label is not processed",
+			inputCR:        azureMachine().withClusterNameLabel(),
+			ownerCluster:   cluster().withReleaseVersionLabel(capiRelease).object,
+			expectedResult: false,
+		},
+		{
+			name:           "AzureMachine with cluster ID label is not processed",
+			inputCR:        azureMachine().withClusterIDLabel(),
+			ownerCluster:   cluster().withReleaseVersionLabel(capiRelease).object,
+			expectedResult: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Log(tc.name)
+			ctx := context.Background()
+			ctrlClient := newFakeClient()
+			loadReleases(t, ctrlClient)
+
+			if tc.ownerCluster != nil {
+				err := ctrlClient.Create(ctx, tc.ownerCluster)
+				if err != nil {
+					t.Fatalf("Error when setting up test data: %#v", err)
+				}
+			}
+
+			result, err := IsObjectReconciledByLegacyRelease(ctx, ctrlClient, ctrlClient, tc.inputCR)
```

Comment:
@tuommaki I passed in `client.Client` here for the reader as well, since here we are testing the filtering logic and not the caching logic.